plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,945 @@
+        An explicitly set time step in units convertible to seconds.
+        Setting this optional keyword overrules the adaptive time step
+        capability and forces the use of this time step throughout.
+
+    dt_range : tuple of shape (2,) of `~astropy.units.Quantity`, optional
+        If specified, the calculated adaptive time step will be clamped
+        between the first and second values.
+
+        field_weighting : str
+        String that selects the field weighting algorithm used to determine
+        what fields are felt by the particles. Options are:
+
+        * 'nearest neighbor': Particles are assigned the fields on
+            the grid vertex closest to them.
+        * 'volume averaged' : The fields experienced by a particle are a
+            volume-average of the eight grid points surrounding them.
+
+        The default is 'volume averaged'.
+
+    req_quantities : `list` of str, optional
+        A list of quantity keys required to be specified on the Grid object.
+        The base particle pushing simulation requires the quantities [E_x, E_y, E_z, B_x, B_y, B_z].
+        If any of these quantities are missing, a warning will be given and that
+        quantity will be assumed to be zero everywhere. This keyword is for specifying quantities in
+        addition to these six. The default is None.
+
+    verbose : bool, optional
+        If true, updates on the status of the program will be printed
+        into the standard output while running. The default is True.
+    """
+
+    def __init__(
+        self,
+        grids: Union[AbstractGrid, Iterable[AbstractGrid]],
+        termination_condition: AbstractTerminationCondition = None,
+        save_routine: Optional[AbstractSaveRoutine] = None,
+        dt=None,
+        dt_range=None,
+        field_weighting="volume averaged",
+        req_quantities=None,
+        verbose=True,
+    ):
+        # self.grid is the grid object
+        if isinstance(grids, AbstractGrid):
+            self.grids = [
+                grids,
+            ]
+        elif isinstance(grids, collections.abc.Iterable):
+            self.grids = grids
+        elif isinstance(grids, BasePlasma):
+            raise TypeError(
+                "It appears you may be trying to access an older version of the ParticleTracker class."
+                "This class has been deprecated."
+                "Please revert to PlasmaPy version 2023.5.1 to use this version of ParticleTracker."
+            )
+        else:
+            raise TypeError("Type of argument `grids` not recognized.")
+
+        # self.grid_arr is the grid positions in si units. This is created here
+        # so that it isn't continuously called later
+        self.grids_arr = [grid.grid.to(u.m).value for grid in self.grids]
+
+        self.verbose = verbose
+
+        # This flag records whether the simulation has been run
+        self._has_run = False
+
+        self._is_adaptive_time_step = False
```

Comment:
It seems like these should be determined by some logic, not just set to False? Presumably down where you have self.dt and self.dt_range being set

Reply:
Yeah the logic for this is currently in `_validate_constructor_inputs` because they're used to determine whether an error should be raised based on the provided save routine and termination condition. Should I move them to the constructor body for the sake of organization?

Reply:
I wouldn't have them both places - having them be created in `validate...` is a bit confusing since I wouldn't expect validation to create new variables. Maybe a new small method `_determine_timestep_type` or something that just does this, then you can call that in the constructor after you validate the input? 

It seems like a lot of methods, but if the names are good that can make the code self-documenting and make it easier to find relevant parts of the code.